pkg/natdiscovery/remote_endpoint.go

```diff
@@ -0,0 +1,138 @@
+/*
+© 2021 Red Hat, Inc. and others.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package natdiscovery
+
+import (
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/submariner-io/submariner/pkg/types"
+)
+
+type endpointState int
+
+const (
+	testingPrivateAndPublicIPs endpointState = iota
+	waitingForResponse
+	selectedPublicIP
+	selectedPrivateIP
+	recheckTime                    = 2 * time.Second
+	totalTimeout                   = 60 * time.Second
+	publicToPrivateFailoverTimeout = 1 * time.Second
+)
+
+type remoteEndpointNAT struct {
+	Endpoint               *types.SubmarinerEndpoint
+	state                  endpointState
+	lastCheck              time.Time
+	lastTransition         time.Time
+	started                time.Time
+	UseNAT                 bool
+	UseIP                  string
+	lastPublicIPRequestID  uint64
+	lastPrivateIPRequestID uint64
+}
+
+func newRemoteEndpointNAT(endpoint *types.SubmarinerEndpoint) *remoteEndpointNAT {
+	return &remoteEndpointNAT{
+		Endpoint:       endpoint,
+		state:          testingPrivateAndPublicIPs,
+		started:        time.Now(),
+		lastTransition: time.Now(),
+	}
+}
+
+func (rnat *remoteEndpointNAT) transitionToState(newState endpointState) {
+	rnat.lastTransition = time.Now()
+	rnat.state = newState
+}
+
+func (rnat *remoteEndpointNAT) sinceLastTransition() time.Duration {
+	return time.Since(rnat.lastTransition)
+}
+
+func (rnat *remoteEndpointNAT) hasTimedOut() bool {
+	return time.Since(rnat.started) > totalTimeout
+}
+
+func (rnat *remoteEndpointNAT) useLegacyNATSettings() {
+	rnat.UseNAT = rnat.Endpoint.Spec.NATEnabled
+	if rnat.Endpoint.Spec.NATEnabled {
+		rnat.UseIP = rnat.Endpoint.Spec.PublicIP
+		rnat.transitionToState(selectedPublicIP)
+	} else {
+		rnat.UseIP = rnat.Endpoint.Spec.PrivateIP
+		rnat.transitionToState(selectedPrivateIP)
+	}
+}
+
+func (rnat *remoteEndpointNAT) shouldCheck() bool {
+	switch rnat.state {
+	case testingPrivateAndPublicIPs:
+		return true
+	case waitingForResponse:
+		return time.Since(rnat.lastCheck) > recheckTime
+	default:
+		return false
+	}
+}
+
+func (rnat *remoteEndpointNAT) checkSent() {
+	if rnat.state == testingPrivateAndPublicIPs {
+		rnat.state = waitingForResponse
+	}
+
+	rnat.lastCheck = time.Now()
+}
+
+func (rnat *remoteEndpointNAT) transitionToPublicIP(remoteEndpointID string, useNAT bool) error {
+	if rnat.state == waitingForResponse {
+		rnat.UseIP = rnat.Endpoint.Spec.PublicIP
+		rnat.transitionToState(selectedPublicIP)
+		rnat.UseNAT = useNAT
```

Comment:
Just for info: Since we are using publicIP here, will useNAT be false always?

Reply:
useNAT will be dynamic, if really the source IP or source Port changed, Imagine of the case where you have a public IP on the interface, and no nat happens.

We haven't seen it, but we would make it compatible this way.